csfieldguide/chapters/content/en/coding-compression/sections/general-purpose.md

```diff
@@ -43,27 +23,308 @@ Here are some of the bits from the example earlier in this chapter; you can past
+
+And this is the compressed message:
+```
+000000000000 000000000001 000000000010 000000000011 000000000101 000000000111 000000000010 000000000100 000000000010
+```
+
+The compressed version shown here uses only 9 LZW codes, but each is 12 bits each, so this version would used 108 bits, which is bigger than the original, although we'll see later on (check out the "Extra for experts: fewer than 12 bits" panel at the end of this section) that there's an easy trick that can be used to reduce this to just 3 or 4 bits per code, so it would use no more than 4x9, or 36 bits, which is less than 40% of the original size.
+
+We were encoding quite a short message, so the longest sequence in the dictionary is just three characters. The longer the message that we are encoding, the longer these sequences can get, therefore allowing us to potentially represent reasonably long substrings with just one code, therefore resulting in effective compression. You can see this for yourself in the LZW compression interactive below:
+
+{interactive slug="lzw-compression" parameters="message-length=long" type="whole-page"}
+
+{panel type="challenge"}
+
+# Challenge
+
+This compression algorithm works best for text with plenty of repeated sequences. To see this for yourself, try encoding the following string using the LZW interactive:
+
+```
+The fox jumped over the lazy dog.
+```
+
+While this string is longer than the one used in our example, there are fewer repeated sequences and therefore not nearly as much opportunity for compression!
+
+{panel end}
+
+
+## LZW Decoding
+
+Decoding a message from LZW is a little tricky, but makes an interesting puzzle. The main point is that anything that has been encoded can be decoded to exactly the same as it was before it was compressed, but here's an illustration of how it works in case you're curious about how it's even possible. Let’s now try to decode our encoded message.
+
+The decoder starts with the same initial dictionary of single characters as the encoder:
+
+| Code | Bits         | Character String |
+|------|--------------|------------------|
+| 0    | 000000000000 | I                |
+| 1    | 000000000001 | A                |
+| 2    | 000000000010 | M                |
+| 3    | 000000000011 | S                |
+
+The encoded message is as follows (we've shown each 12-bit code as a decimal number):
+
+```
+0 1 2 3 5 7 2 4 2
+```
+
+The initial dictionary has values for codes `0 - 3`, so what is the decoder meant to do when it gets to `4`? Well, at each step it makes an addition to the dictionary, just as the encoder did. We’ll walk through this step by step:
+
+| Step | Code | Output | New Dict Entry Code | New Dict Sequence | Next Proposed Dict Sequence |
+|------|------|--------|---------------------|-------------------|-----------------------------|
+| 1    | 0    | I      | -                   | -                 | I?                          |
```

Comment:
Is there a way to make the heading look nicer?
![image](https://user-images.githubusercontent.com/25916475/55844330-e4968a00-5b90-11e9-8063-844bcbae05bd.png)

Reply:
I think there's room for a whole other PR here for improving the styling of tables.

Reply:
How about `New Code`, `New Sequence` and `Proposed Sequence` or similar so the words don't wrap?